### PR TITLE
Pagination component (closes #3, closes https://github.com/spatie/smartassist/issues/221)

### LINKED ISSUE #3: Add Pagination component
- Deprecate the `paginator` slot scope prop
- Expose a `pageCount` slot scope prop instead (`Math.ceil(totalCount / pageSize`)

https://github.com/spatie/vue-data-component/blob/eefd8371192bfa5a685d0ad5b5d91f5b513916cc/src/components/DataComponent.js#L184

- Add a new `data-paginator` renderless component. `pages` slot scope prop contains an object like the old `paginator`

```vue
<data-paginator :page="page" :page-count="pageCount">
    <ul slot-scope="{ pages }">
        <li v-for="page in pages" :class="{ 'is-active': page.active }">
            {{ page.number }}
        </li>
    </ul>
</data-paginator>
```

- Add an optional `:links-on-each-side` prop to `data-paginator` to reduce the number of pages in the `pages` slot scope prop (idea stolen from https://laravel-news.com/laravel-5-7-pagination-link-customizations)

```vue
<data-paginator :page="page" :page-count="pageCount" :links-on-each-side="2">
    <ul slot-scope="{ pages, firstPage, lastPage }">
        <li :class="{ 'is-active': firstPage.active }">
            {{ firstPage.number }}
        </li>
        <li>...</li>
        <li v-for="page in pages" :class="{ 'is-active': page.active }">
            {{ page.number }}
        </li>
        <li>...</li>
        <li :class="{ 'is-active': lastPage.active }">
            {{ lastPage.number }}
        </li>
    </ul>
</data-paginator>
```

The last example is going to be a bit harder to implement. Some example (edge) cases:

```
<data-paginator :page="4" :page-count="10" :links-on-each-side="2">
1 ... 2 3 [4] 5 6 ... 10

<data-paginator :page="9" :page-count="10" :links-on-each-side="2">
1 ... 7 8 [9] 10

<data-paginator :page="2" :page-count="3" :links-on-each-side="2">
1 [2] 3
```

- Other and better ideas are more than welcome 😄

### PR DESCRIPTION
**Pull request by AdrianMrn**

### CODE REVIEW

File: docs/pages/ajax-card-layout.vue [hunk 1]
```diff
@@ -8,8 +8,9 @@
             :fetcher="getRicks"
             :query.sync="query"
             :initial-load-delay-ms="400"
+            pageCountKey="pageCount"
```
Comment (sebastiandedeyne) [inline docs/pages/ajax-card-layout.vue]: Kebab case :D

File: src/components/DataComponent.js [hunk 2]
```diff
@@ -17,7 +16,7 @@ export default {
         pageNumberKey: { default: 'page' },
         pageSizeKey: { default: 'pageSize' },
         pageCountKey: { default: null },
```
Comment (sebastiandedeyne) [inline src/components/DataComponent.js]: You can add a `pageCount` default here (and drop the explicitly passed prop from the previous comment)

File: src/components/DataPaginator.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+            // https://gist.github.com/kottenator/9d936eb3e4e3c3e02598#gistcomment-1748957
+            let range = []
+
+            for (let i = Math.max(2, this.page - this.linksOnEachSide); i <= Math.min(this.pageCount - 1, this.page + this.linksOnEachSide); i++) {
+                range.push(i);
+            }
+
+            if (this.page - this.linksOnEachSide > 2) {
+                range.unshift("...");
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: `...` -> `…`

File: src/components/DataPaginator.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                pages: this.paginator,
+            });
+        }
+
+        if (this.pageCount === 1) {
+            return null;
+        }
+
+        return (
+            <ul slot-scope="{ pages }" class="mt-4 flex justify-center">
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: This `slot-scope` can go

File: src/components/DataPaginator.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                                onClick={() => this.$emit('change-page', page.number)}
+                                disabled={ page.disabled }
+                            >
+                                { page.number }
+                            </button>
+                        }
+                    </li>
+                ) }
+                <button
+                    onClick={() => this.$emit('change-page', (this.page === this.pageCount) ? this.page : this.page + 1)}
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: Move this to it's own method and use it at line 71 too?

Also, keep even names like html does -> no "spaces", first subject then action. -> `pagechange`

File: src/components/DataPaginator.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                <li>
+                    <button
+                        onClick={() => this.$emit('change-page', (this.page === 1) ? this.page : this.page - 1)}
+                    >
+                        &lt;
+                    </button>
+                </li>
+                { this.paginator.map(page =>
+                    <li key={ page.number } class={page.active ? 'active' : ''}>
+                        { page.disabled &&
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: Maybe use a ternary operator instead?

```js
{page.disabled ? (
    page.number
) : (
    <button></button>
)}
```

File: src/components/DataPaginator.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            });
+        }
+
+        if (this.pageCount === 1) {
+            return null;
+        }
+
+        return (
+            <ul slot-scope="{ pages }" class="mt-4 flex justify-center">
+                <li>
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: These should get a disabled state and not be a button when we're on the first/last page

File: src/components/DataPaginator.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    props: {
+        page: { required: true },
+        pageCount: { required: true, type: Number },
+        linksOnEachSide: { default: 2, type: Number },
+        navigateButtons: { default: true, type: Boolean},
+    },
+
+    computed: {
+        paginator() {
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: Since this is just an array of page objects, should we call it `pages` instead?

**Review by sebastiandedeyne — CHANGES_REQUESTED** [review]

File: src/components/DataPaginator.js (hunk 8 above)
Comment (AdrianMrn) [reply src/components/DataPaginator.js]: 👌

File: src/components/DataPaginator.js (hunk 6 above)
Comment (AdrianMrn) [reply src/components/DataPaginator.js]: 👍

File: src/components/DataComponent.js (hunk 2 above)
Comment (AdrianMrn) [reply src/components/DataComponent.js]: @sebastiandedeyne https://github.com/spatie/vue-data-component/blob/b0ce1bb23a2e186596f48af54bf3a5f622ef4f9d/src/components/DataComponent.js#L226-L232
Line 231 wouldn't work anymore (for local data sources)

File: src/components/DataPaginator.js (hunk 7 above)
Comment (AdrianMrn) [reply src/components/DataPaginator.js]: @sebastiandedeyne Like this?
```jsx
<li>
  { this.page === this.pageCount ? '>' : (
    <button
      onClick={() => this.pageChange((this.page === this.pageCount) ? this.page : this.page + 1)}
    >
      &gt;
    </button>
  )}
</li>
```

File: src/components/DataComponent.js (hunk 2 above)
Comment (sebastiandedeyne) [reply src/components/DataComponent.js]: We can work around this by attempting to retrieve the page count, and fall back on our calculation if nothing was returned.

```js
const pageCount = get(fetchResult, this.pageCountKey);

if (pageCount === undefined) { 
    return pageCount;
}
```

Remember, `0` is a valid page count!

File: src/components/DataPaginator.js (hunk 7 above)
Comment (sebastiandedeyne) [reply src/components/DataPaginator.js]: On second though, maybe keep the `button` element here with `disabled` 😅 I think it's more semantic like that

**Comment by sebastiandedeyne** [comment]
One more general comment: could you look into aria attributes and see if they have anything regarding pagination that we could/should add?

**Comment by AdrianMrn** [comment]
> One more general comment: could you look into aria attributes and see if they have anything regarding pagination that we could/should add?

Done! Used http://www.a11ymatters.com/pattern/pagination/ and https://www.w3.org/TR/html5/dom.html#do-not-set

@sebastiandedeyne Merge if satisfied :)

File: docs/pages/ajax-card-layout.vue [hunk 9]
```diff
@@ -118,9 +109,10 @@ export default {
         getRicks({ query, queryString }) {
             const baseUrl = 'https://rickandmortyapi.com/api/character';
 
-            return axios.get(`${baseUrl}?name=Rick&${queryString}`).then(response => ({
+            return axios.get(`${baseUrl}?${queryString}`).then(response => ({
```
Comment (sebastiandedeyne) [inline docs/pages/ajax-card-layout.vue:112]: This can stay, we still only care about Rick :)

File: src/components/DataPaginator.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        if (this.pageCount === 1) {
+            return null;
+        }
+
+        return (
+            <nav aria-label="Pagination Navigation">
+                <ul class="mt-4 flex justify-center">
+                    <li>
+                        <button
+                            onClick={ () => this.pageChange((this.page === 1) ? this.page : this.page - 1) }
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js]: Can you run `yarn format` before merging? I think prettier should strip the spaces after `{` and before `}`

File: docs/pages/ajax-card-layout.vue [hunk 11]
```diff
@@ -64,17 +65,8 @@
...
-                        <button
-                            class="mx-4"
-                            :class="page.active ? 'border-b border-black' : 'text-grey-dark'"
-                            @click="query.page = page.number"
-                        >
-                            {{ page.number }}
-                        </button>
-                    </li>
-                </ul>
+                <data-paginator :page="query.page" :page-count="pageCount" @page-change="query.page = $event"></data-paginator>
```
Comment (sebastiandedeyne) [inline docs/pages/ajax-card-layout.vue:68]: `page-change` shouldn't exist, it should be an `input` event so `page` can be bound with `v-model`

File: src/components/DataPaginator.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                pages: this.pages,
+            });
+        }
+
+        if (this.pageCount === 1) {
+            return null;
+        }
+
+        return (
+            <nav aria-label="Pagination Navigation">
```
Comment (sebastiandedeyne) [inline src/components/DataPaginator.js:66]: You're going to need to install the necessary Babel plugins to support JSX (and run `yarn build` before to ensure it all transpiles well)

https://github.com/vuejs/babel-plugin-transform-vue-jsx

**Review by sebastiandedeyne — CHANGES_REQUESTED** [review]

File: docs/pages/ajax-card-layout.vue (hunk 11 above)
Comment (AdrianMrn) [reply docs/pages/ajax-card-layout.vue:68]: slim!

File: src/components/DataPaginator.js (hunk 12 above)
Comment (AdrianMrn) [reply src/components/DataPaginator.js:66]: installed plugins & added them to `.babelrc` but I must've done something wrong, it's not working yet :#

**Review by sebastiandedeyne — APPROVED** [review]

File: src/components/DataPaginator.js (hunk 12 above)
Comment (sebastiandedeyne) [reply src/components/DataPaginator.js:66]: Apparently the Rollup Babel plugin must come before the Rollup Commonjs plugin

**Pull request merged**